XboxHelper.py

```diff
@@ -289,21 +294,24 @@ def print_cache_state(self):
         print("Put / Pusher enabled: %s" % ("Yes" if (push_state & 1) else "No"))
         print("Get / Puller enabled: %s" % ("Yes" if (pull_state & 1) else "No"))
 
-        print("Cache:")
-        for i in range(128):
+        if print_contents:
+            print("Cache:")
+            # Each CACHE entry is a pair of 32-bit integers and there are 128 entries.
+            methods_and_data = self.xbox.read(CACHE1_METHOD, 8 * 128)
```

Comment:
This is risky, because it depends on the implementation of the `read`. You are accessing MMIO here.
I believe I specified xboxpy to load 32 bit chunks first, then 16 bit and finally 8 bit towards the end of the buffer.
So this *should* work, but another backend in xboxpy might still get it wrong, and a naive port to another backend would also fail.

I think it's better to keep this somewhat specific at least (so consider making a self-documenting high-level wrapper like `read_multiple_u32` / `read_u32(..., count=X)` or similar).

The main `read` operation in xboxpy might be optimized in the future. I had always had plans to enable some form of page-mirroring which actually mirrors the page-table and installs a watchdog to look for changes, so only touched changes would be transferred again, with the client keeping a mirror of the memory. Additionally, I had always planned to add compression for larger transfers, or DMA to blast the pages directly via ethernet (without lwip interaction).
xboxpy basically needs a way to differentiate MMIO and bulk-data-read/write.

The performance gain of this could be massive.

Reply:
**Reading multiple u32**

I think a smarter protocol which batches multiple operation would be useful.
Ideally there'd be something which disables interrupts, disables memory protection, then runs a number of queued commands, and only then enables protection and interrupts again.

Basically what we'd need is a transaction system in xboxpy, so you can batch multiple actions together.

**Memory mirroring to optimize bulk reads**

I'm not sure where to put some memory-mirroring system either. In theory, nothing prevents it from being independent of the xboxpy backend, so it could live in nv2a-trace.
The tricky bit is that it would probably have to install an interrupt handler.

This could be implemented in xboxpy, or in nv2a-trace, although I'd prefer it to live in nv2a-trace, because that's probably the only user.

---

The backends like XBDM and nxdk-rdt were meant for tiny atomic operations. The macro operations in nxdk-rdt or xbdm are useful for humans interacting with the Xbox, but they are pretty useless most of the time, while adding a lot of complexity.

Originally, xboxpy was meant to provide a high level view of the Xbox (on top of the atomic operations in the backends). Therefore xboxpy had an API for the APU and NV2A. However, it led to unnecessary complexity and most of the time there was a single user for those device specific APIs and then those APIs were often too complicated.
Instead, I feel like it was much more useful to be able to inject and call code manually from the actual applications / scripts.

So if I had to write xboxpy from scratch, I'd make it transaction based and focus on having a simple interface for MMIO, reads/writes and calls, potentially with the option to load C code (although I'd probably write the binary loader in Python on the PC side).

But I digress again.

Reply:
I think this is very useful context and certainly helps me get an idea of where it might make sense to put in further effort. From your other comments below I think it's clear that moving more of the responsibility for tracing onto the hardware would be a good design direction generally. I think it would also reduce the testing surface (i.e., remove worries about behavior dependent on specific xboxpy interfaces) and having the general ability to inject code is obviously useful in a variety of situations, becoming even more useful if the bar is lowered by allowing C in addition to asm.